### DISCUSSION TITLE
"wifiSleepTime" for energy saving and reducing wifi smog

### DISCUSSION THREAD

**Discussion by Gifford47** [arendst/Tasmota#11459] [Ideas]
i currently have a d1 mini with a bme260 sensor connected to it. now i had tried the deepsleep and noticed that every time i reconnect to the wifi, a lot of status data is sent and also the stored values var1,var2, etc. are lost.

since the esp consumes a relatively large amount of power in online mode, it would make sense if i could switch to offline mode by rule, for example, and still be able to evaluate all calculations or sensors.

I am aware that some rules cannot be executed in offline mode, but this can be avoided if the esp has been online once and has saved the time in the rtc. then it can switch off the wifi with a new command "wifiSleepTime", for example (similar to DeepSleepTime). after the time (wifiSleepTime) has elapsed, it reconnects and looks for new commands or sends the tele status. After that it goes to offline mode.
By rule, i then still have the choice of connecting to the wifi before the wifiSleepTime expires and sending a status.

this has already been requested several times in the issues.

**Comment by barbudor** [comment]
If you are talking about ModemSleep mode, this is already in use
See command [`Sleep`](https://tasmota.github.io/docs/Commands/#sleep)

**Reply by Gifford47** [reply]
no, I would like to cut the wifi connection completely for a longer period of time. preferably with a function that I can execute during booting so that I can switch the wifi off and on if necessary.

**Reply by meingraham** [reply]
This topic has been discussed many times before and the outcome has never wavered - unfortunately for your request, it will most likely still not be implemented in Tasmota for Wi-Fi only ESP-based modules. Turning off Wi-Fi is only supported for modules that have an Ethernet port onboard and this is currently available for Tasmota32 only.

You do have the option to create your own fork and modify, maintain, and compile the code yourself.

**Reply by Gifford47** [reply]
i have read the threads, but unfortunately i never understood the point behind them, as the module could also be used offline in many ways with tasmota. the wifi connection is not the main feature of the module in my opinion.
it is a sad fact that the user is not given the choice of using the module online or offline.

**Reply by meingraham** [reply]
Absolutely! The user of this **_FREE_** Tasmota software has the option. You can **freely** make the modification yourself.

These Espressif SoC are **Wi-Fi** SoC for use in Wi-Fi applications. Thus, the Wi-Fi connection **IS** the main feature.

**Reply by Gifford47** [reply]
ok, i gladly accept other opinions and i'm also glad that the software is opensource. however, this misses the issue.

is there anything that would speak against including the option in my_user_config.h?

**Reply by meingraham** [reply]
wifiSleepTime would be new code for a new command so not just a compile time option. Yes, wifiSleepTime could be a compile time option... so someone with the knowledge and time to code the new command can do so and submit a PR. This is a matter of the effort, but also the additional flash and RAM required. These are usually the factors that determine whether new features are incorporated into Tasmota.

As for turning off Wi-Fi, that command is available in Tasmota32 but not in Tasmota. Again, your own fork could make this command available in Tasmota. However, there is the question as to how reliable ESP82xx operation is if this code is compiled and used. Perhaps you could test that and provide feedback to the rest of the Tasmota user community.

**Reply by Gifford47** [reply]
i just don't understand why the feature is not included because of the reason: the node would not be longer a smart home device. If i want to put my node to sleep using deepsleep to save energy, i think it is the same effect as switching off the wifi. both are offline and save energy. however, with "wifioff" i can still read sensors and switch on the wifi when i need it. for me, this would be a big advantage of a smart home device...
please don't get me wrong, but i think this feature would be interesting several users

**Comment by Gifford47** [comment]
i have now started working on a fork and testing the "wifi off" command in a first binary. unfortunately i have limited experience in the tasmota branch so i hope someone will help me out

**Comment by Gifford47** [comment]
i have successfully got the first version running. build is tasmota 9.3.1 (sensors). everything works fine for me at the moment. the "new" command is "wifi on" and "wifi off".

could someone also test this build?

[tasmota-sensors9.3.1.bin.gz](https://github.com/arendst/Tasmota/files/6228624/tasmota-sensors9.3.1.bin.gz)

**Reply by barbudor** [reply]
This command will conflict with existing `Wifi` command for ESP32 (to turn wifi off when using Ethernet)

Can you please share your fork/branch instead of a binary ?
What is your actual current saving measurements ?

Thanks

**Reply by Gifford47** [reply]
I actually just used the code and command of the esp32 and took the function from the #ifdef esp32 as a test. Unfortunately I could not understand why there will be a conflict with the esp32? can you please explain it to me?

https://github.com/Gifford47/Tasmota

**Reply by barbudor** [reply]
If you just enabled the existing code for esp8266 there should be no problem.
I meant that we probably don't want to have a function CmndWifi somewhere for ESP32 and somewhere else for esp8266, or the same command having different effect on ESP32 and esp8266. 

I can try to test your fork this long weekend. Thanks for sharing

**Reply by sfromis** [reply]
At some point, the wifi disable code was accidentally included for ESP8266. As this proved bad, it was excluded.

**Reply by Gifford47** [reply]
ok, i have found two "errors":
1.when i turn off the wifi it restarts and loses the time. so it must work without restarting.
2. the esp starts an AP when the wifi is switched off. of course, this does not save any energy...

**Comment by Gifford47** [comment]
ok, i changed the code again and now everything works great, you can turn wifi on and off without rebooting. however, the esp tries to reconnect to the mqtt broker when wifi off. unfortunately, i could not use the wifi_shutdown function... did you test my code?

**Reply by barbudor** [reply]
I didn't as you said it was not stable.
I will have a look may be during the week
But as I would like to make some current measurement too that will not before the week end.
Thanks

**Reply by Gifford47** [reply]
Hey!
Do you have any results?
so far everything is running fine for me. except that i can't turn off the automatic reconnect from mqtt.... do you have any ideas?

**Reply by Gifford47** [reply]
so, i have an update:
i measured the current consumption of one esp 8266 (wemos D1 mini). with wifi off, the esp consumes 24.5 mA on average (measurement time 60s). with wifi on, 43.5 mA on average (measurement time 60s).
the wifi can be switched on and off during operation. the mqtt service connects automatically on reconnect. in addition, the time is automatically synchronised once the esp has been on the internet.

my last branch (development) has the latest firmware...
can anyone confirm these results for the esp32?

**Reply by sfromis** [reply]
Did you measurement include the extra power needed for each reconnect?

**Reply by ortegafernando** [reply]
Hi, could you please, write an example (rules, sensors....)  and write it also into tasmota/docs ?

**Reply by Gifford47** [reply]
> Did you measurement include the extra power needed for each reconnect?

no, i only measured the current if the module is already connected or disconnected. because in case of reconnect there are short peaks up to 85 ma which affect the average value.

**Reply by sfromis** [reply]
It is exactly because those peaks will affect the average that it is important to include them when you go on about "energy saving" (likely tiny). There will be a point when reconnections burns more power than just keeping wifi on.
